File: linearlist/LRUCache146.cc
/**
 * 146. LRU 缓存机制
 * 
 */
#include<unordered_map>
using namespace std;
typedef struct Node{
    int key,val;
    Node *next, *prev;
    Node(int k,int v):key(k), val(v) ,next(nullptr),prev(nullptr) {};
}Node;

class DoubleList{
    Node *head, *tail;
    //链表元素数
    int size;
public:
    void initDoubleList(){
        head = new Node(0,0);
        tail = new Node(0,0);
        head->next = tail;
        tail->prev = head;
        size = 0;
    }
    //在链表尾部添加节点
    void addLast(Node *x){
        x->prev = tail->prev;
        x->next = tail;
        tail->prev->next = x;
        tail->prev = x;
        size++;
    }  
    //删除链表中的x节点
    void remove(Node *x){
        x->prev->next = x->next;
        x->next->prev = x->prev;
        size--;
    }
    //删除链表中到第一个节点
    Node* removeFirst(){
        if(head->next == tail) return nullptr;
        Node *first = head->next;
        remove(first);
        return first;
    }

    int getSize(){
        return size;
    }
};
class LRUCache {
public:
    //key -> Node(key,val)
    unordered_map<int, Node*> map;
    //lined  最后一位是最近使用  第一位是使用最久
    DoubleList cache;
    int cap ;
    LRUCache(int capacity) {
        cache =  DoubleList();
        cache.initDoubleList();
        cap = capacity;
    }
    //将某个key提升为最近使用
    void useRecently(int k){
        auto iterator = map.find(k);
        cache.remove(iterator->second);
        cache.addLast(iterator->second);
    }
    //添加到最近使用
    void addRecently(int key, int val){
        Node *x = new Node(key,val);
        cache.addLast(x);
        map[key] = x;
    }
    //删除某一个key
    void deletekey(int key){
        auto x = map.find(key);
        cache.remove(x->second);
        map.erase(x);
    }
    //删除较长时间未使用
    void removeLeastRecently(){
        Node *firstNode = cache.removeFirst();
        int deleteKey = firstNode->key;
        map.erase(deleteKey);
    }

    
    int get(int key) {
        if(map.find(key) == map.end()) return -1;
        useRecently(key);
        return map[key]->val;
    }
    
    void put(int key, int value) {
        if(map.find(key) != map.end()){
            deletekey(key);
            addRecently(key,value);
            return;
        }
        if(cap == cache.getSize()) removeLeastRecently();
        addRecently(key, value);
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: linearlist/lrucache146/lrucache146.go
package lrucache146

//DListNode 链表的结点
type DListNode struct {
	key  int
	val  int
	prio *DListNode
	next *DListNode
}

//LRUCache 双向链表
type LRUCache struct {
	cache    map[int]*DListNode
	capacity int
	head     *DListNode
	tail     *DListNode
}

//Constructor 构造LRUCache 双向链表
func Constructor(capacity int) LRUCache {
	lru := LRUCache{
		cache:    map[int]*DListNode{},
		capacity: capacity,
		head:     &DListNode{key: 0, val: 0},
		tail:     &DListNode{key: 0, val: 0},
	}
	lru.head.next = lru.tail
	lru.tail.prio = lru.head
	return lru
}

//RemoveNode 删除制定结点
func (l *LRUCache) RemoveNode(node *DListNode) {
	node.prio.next = node.next
	node.next.prio = node.prio
}

//RemoveNode 删除最后一个结点
func (l *LRUCache) RemoveNodeLast() *DListNode {
	lastNode := l.tail.prio
	l.RemoveNode(lastNode)
	return lastNode
}

//AddNodeFirst 添加结点到第一位
func (l *LRUCache) AddNodeFirst(node *DListNode) {
	node.prio = l.head
	node.next = l.head.next
	l.head.next.prio = node
	l.head.next = node
}

//MoveNodeFirst 移动结点到第一位
func (l *LRUCache) MoveNodeFirst(node *DListNode) {
	l.RemoveNode(node)
	l.AddNodeFirst(node)
}

//Get
// 1. 如果获取的到 则将缓存放在第一位
// 2.  获取不到 返回-1
func (l *LRUCache) Get(key int) int {
	if v, ok := l.cache[key]; ok {
		l.MoveNodeFirst(v)
		return v.val
	}
	return -1
}

//Put
// 1. key存在 则将缓存放在第一位，并且更新value
// 2. key不存在 则创建新的链表结点放到第一位，并且添加到cache中
// 3. 缓存满了 删除最后一位 并且删除cache中的key
func (l *LRUCache) Put(key int, value int) {
	node := l.cache[key]
	if node != nil {
		node.val = value
		l.MoveNodeFirst(node)
		return
	}
	node = &DListNode{key: key, val: value}
	l.AddNodeFirst(node)
	l.cache[key] = node

	if len(l.cache) > l.capacity {
		rmNode := l.RemoveNodeLast()
		delete(l.cache, rmNode.key)
	}

}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
